File: src/main/java/io/github/takenoha/towerdefense/paper/TacticalBuildSelectionGui.java
package io.github.takenoha.towerdefense.paper;

import io.github.takenoha.towerdefense.tactical.TacticalBuildDefinition;
import io.github.takenoha.towerdefense.tactical.TacticalCandidate;
import io.github.takenoha.towerdefense.tactical.TacticalCandidateSet;
import java.util.List;
import java.util.Objects;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;

/** Small, one-screen candidate selection UI used before the existing start transaction. */
public final class TacticalBuildSelectionGui {
    public static final int SIZE = 27;
    public static final int[] CANDIDATE_SLOTS = {11, 13, 15};
    public static final int[] BRANCH_SLOTS = {3, 5};
    public static final int CONFIRM_SLOT = 22;
    public static final int CLOSE_SLOT = 26;

    private TacticalBuildSelectionGui() {
    }

    public static Inventory create(TacticalBuildSelectionInventoryHolder holder) {
        Objects.requireNonNull(holder, "holder");
        Inventory inventory = Bukkit.createInventory(
                holder,
                SIZE,
                Component.text("戦術ビルドを選択", NamedTextColor.DARK_PURPLE));
        holder.attach(inventory);
        refresh(
                inventory,
                holder.candidates(),
                holder.selectedBuildId().orElse(null),
                holder.selectedBranchId().orElse(null));
        inventory.setItem(
                CONFIRM_SLOT,
                item(
                        Material.LIME_CONCRETE,
                        "このビルドで開始",
                        List.of("選択後にクリックすると開始確認へ進みます。"),
                        NamedTextColor.GREEN));
        inventory.setItem(
                CLOSE_SLOT,
                item(
                        Material.BARRIER,
                        "キャンセル",
                        List.of("閉じても開始印は消費されません。"),
                        NamedTextColor.RED));
        return inventory;
    }

    public static void refresh(
            Inventory inventory,
            TacticalCandidateSet candidates,
            String selectedBuildId) {
        refresh(inventory, candidates, selectedBuildId, null);
    }

    public static void refresh(
            Inventory inventory,
            TacticalCandidateSet candidates,
            String selectedBuildId,
            String selectedBranchId) {
        for (int branchSlot : BRANCH_SLOTS) {
            inventory.setItem(branchSlot, null);
        }
        for (int slot = 0; slot < CANDIDATE_SLOTS.length; slot++) {
            TacticalBuildDefinition definition = candidates.candidates().get(slot).definition();
            boolean selected = definition.id().equals(selectedBuildId);
            inventory.setItem(
                    CANDIDATE_SLOTS[slot],
                    item(
                            material(definition.iconMaterial()),
                            (selected ? "▶ " : "") + definition.displayName(),
                            List.of(
                                    definition.description(),
                                    "カテゴリ: " + definition.category(),
                                    "レアリティ: " + definition.rarity(),
                                    selected ? "選択中" : "クリックで選択"),
                            selected ? NamedTextColor.GREEN : NamedTextColor.AQUA));
        }
        if (selectedBuildId == null) {
            return;
        }
        TacticalBuildDefinition selected = candidates.requireBuild(selectedBuildId);
        List<String> branchIds = selected.branchIds();
        for (int index = 0; index < Math.min(branchIds.size(), BRANCH_SLOTS.length); index++) {
            String branchId = branchIds.get(index);
            boolean branchSelected = branchId.equals(selectedBranchId);
            inventory.setItem(
                    BRANCH_SLOTS[index],
                    item(
                            branchMaterial(branchId),
                            (branchSelected ? "▶ " : "") + branchName(branchId),
                            branchLore(selected, branchId, branchSelected),
                            branchSelected ? NamedTextColor.GREEN : NamedTextColor.LIGHT_PURPLE));
        }
    }

    public static int candidateIndexAt(int rawSlot) {
        for (int index = 0; index < CANDIDATE_SLOTS.length; index++) {
            if (CANDIDATE_SLOTS[index] == rawSlot) {
                return index;
            }
        }
        return -1;
    }

    public static int branchIndexAt(int rawSlot) {
        for (int index = 0; index < BRANCH_SLOTS.length; index++) {
            if (BRANCH_SLOTS[index] == rawSlot) {
                return index;
            }
        }
        return -1;
    }

    private static List<String> branchLore(
            TacticalBuildDefinition definition,
            String branchId,
            boolean selected) {
        List<String> lore = new java.util.ArrayList<>();
        lore.add("この枝のTier効果:");
        definition.nodes().stream()
                .filter(node -> node.branchId().filter(branchId::equals).isPresent())
                .sorted(java.util.Comparator.comparingInt(
                        io.github.takenoha.towerdefense.tactical.TacticalSkillNodeDefinition::tier))
                .forEach(node -> lore.add(
                        "Tier " + node.tier() + ": " + node.description()));
        lore.add(selected ? "選択中" : "クリックで選択");
        return lore;
    }

    private static String branchName(String branchId) {
        return switch (branchId) {
            case "rapid-fire" -> "連射ルート";
            case "range" -> "射程ルート";
            default -> branchId + "ルート";
        };
    }

    private static Material branchMaterial(String branchId) {
        return switch (branchId) {
            case "rapid-fire" -> Material.SPECTRAL_ARROW;
            case "range" -> Material.SPYGLASS;
            default -> Material.PAPER;
        };
    }

    private static Material material(String name) {
        Material material = Material.matchMaterial(name);
        return material == null ? Material.PAPER : material;
    }

    private static ItemStack item(
            Material material,
            String name,
            List<String> lore,
            NamedTextColor color) {
        ItemStack item = new ItemStack(material);
        ItemMeta meta = Objects.requireNonNull(item.getItemMeta(), "selection item metadata");
        meta.displayName(Component.text(name, color));
        meta.lore(lore.stream().map(line -> Component.text(line, NamedTextColor.GRAY)).toList());
        meta.addItemFlags(ItemFlag.HIDE_ATTRIBUTES);
        item.setItemMeta(meta);
        return item;
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/paper/TacticalBuildSelectionGui.kt
package io.github.takenoha.towerdefense.paper

import io.github.takenoha.towerdefense.tactical.TacticalBuildDefinition
import io.github.takenoha.towerdefense.tactical.TacticalCandidateSet
import java.util.ArrayList
import java.util.Collections
import java.util.Objects
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.ItemFlag
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.ItemMeta

/** Small, one-screen candidate selection UI used before the existing start transaction. */
class TacticalBuildSelectionGui private constructor() {
    companion object {
        const val SIZE: Int = 27
        @JvmField
        val CANDIDATE_SLOTS: IntArray = intArrayOf(11, 13, 15)
        @JvmField
        val BRANCH_SLOTS: IntArray = intArrayOf(3, 5)
        const val CONFIRM_SLOT: Int = 22
        const val CLOSE_SLOT: Int = 26

        @JvmStatic
        fun create(holder: TacticalBuildSelectionInventoryHolder): Inventory {
            Objects.requireNonNull(holder, "holder")
            val inventory = Bukkit.createInventory(
                holder,
                SIZE,
                Component.text("戦術ビルドを選択", NamedTextColor.DARK_PURPLE),
            )
            holder.attach(inventory)
            refresh(
                inventory,
                holder.candidates(),
                holder.selectedBuildId().orElse(null),
                holder.selectedBranchId().orElse(null),
            )
            inventory.setItem(
                CONFIRM_SLOT,
                item(
                    Material.LIME_CONCRETE,
                    "このビルドで開始",
                    listOf("選択後にクリックすると開始確認へ進みます。"),
                    NamedTextColor.GREEN,
                ),
            )
            inventory.setItem(
                CLOSE_SLOT,
                item(
                    Material.BARRIER,
                    "キャンセル",
                    listOf("閉じても開始印は消費されません。"),
                    NamedTextColor.RED,
                ),
            )
            return inventory
        }

        @JvmStatic
        fun refresh(
            inventory: Inventory,
            candidates: TacticalCandidateSet,
            selectedBuildId: String?,
        ) {
            refresh(inventory, candidates, selectedBuildId, null)
        }

        @JvmStatic
        fun refresh(
            inventory: Inventory,
            candidates: TacticalCandidateSet,
            selectedBuildId: String?,
            selectedBranchId: String?,
        ) {
            for (branchSlot in BRANCH_SLOTS) {
                inventory.setItem(branchSlot, null)
            }
            for (slot in CANDIDATE_SLOTS.indices) {
                val definition = candidates.candidates().get(slot).definition()
                val selected = definition.id() == selectedBuildId
                inventory.setItem(
                    CANDIDATE_SLOTS[slot],
                    item(
                        material(definition.iconMaterial()),
                        (if (selected) "▶ " else "") + definition.displayName(),
                        listOf(
                            definition.description(),
                            "カテゴリ: ${definition.category()}",
                            "レアリティ: ${definition.rarity()}",
                            if (selected) "選択中" else "クリックで選択",
                        ),
                        if (selected) NamedTextColor.GREEN else NamedTextColor.AQUA,
                    ),
                )
            }
            val selectedBuild = selectedBuildId ?: return
            val selected = candidates.requireBuild(selectedBuild)
            val branchIds = selected.branchIds()
            for (index in 0 until minOf(branchIds.size, BRANCH_SLOTS.size)) {
                val branchId = branchIds.get(index)
                val branchSelected = branchId == selectedBranchId
                inventory.setItem(
                    BRANCH_SLOTS[index],
                    item(
                        branchMaterial(branchId),
                        (if (branchSelected) "▶ " else "") + branchName(branchId),
                        branchLore(selected, branchId, branchSelected),
                        if (branchSelected) NamedTextColor.GREEN else NamedTextColor.LIGHT_PURPLE,
                    ),
                )
            }
        }

        @JvmStatic
        fun candidateIndexAt(rawSlot: Int): Int {
            for (index in CANDIDATE_SLOTS.indices) {
                if (CANDIDATE_SLOTS[index] == rawSlot) {
                    return index
                }
            }
            return -1
        }

        @JvmStatic
        fun branchIndexAt(rawSlot: Int): Int {
            for (index in BRANCH_SLOTS.indices) {
                if (BRANCH_SLOTS[index] == rawSlot) {
                    return index
                }
            }
            return -1
        }

        private fun branchLore(
            definition: TacticalBuildDefinition,
            branchId: String,
            selected: Boolean,
        ): List<String> {
            val lore = ArrayList<String>()
            lore.add("この枝のTier効果:")
            definition.nodes().asSequence()
                .filter { node ->
                    node.branchId().filter { value -> branchId == value }.isPresent
                }
                .sortedBy { node -> node.tier() }
                .forEach { node -> lore.add("Tier ${node.tier()}: ${node.description()}") }
            lore.add(if (selected) "選択中" else "クリックで選択")
            return lore
        }

        private fun branchName(branchId: String): String = when (branchId) {
            "rapid-fire" -> "連射ルート"
            "range" -> "射程ルート"
            else -> "${branchId}ルート"
        }

        private fun branchMaterial(branchId: String): Material = when (branchId) {
            "rapid-fire" -> Material.SPECTRAL_ARROW
            "range" -> Material.SPYGLASS
            else -> Material.PAPER
        }

        private fun material(name: String): Material = Material.matchMaterial(name) ?: Material.PAPER

        private fun item(
            material: Material,
            name: String,
            lore: List<String>,
            color: NamedTextColor,
        ): ItemStack {
            val item = ItemStack(material)
            val meta: ItemMeta = Objects.requireNonNull(item.itemMeta, "selection item metadata")
            meta.displayName(Component.text(name, color))
            meta.lore(
                Collections.unmodifiableList(
                    ArrayList(lore.map { line -> Component.text(line, NamedTextColor.GRAY) }),
                ),
            )
            meta.addItemFlags(ItemFlag.HIDE_ATTRIBUTES)
            item.itemMeta = meta
            return item
        }
    }
}
